### PR TITLE
E2E Template from Examples

### PR DESCRIPTION
**Pull request by avishniakov**
This PR is mostly lift&shift of https://github.com/zenml-io/zenml/tree/develop/examples/e2e.
**What was added**
- Testing of various templating scenarios
- Templating itself with certain parametrization (see README for details)
- Minimalistic CI for running tests (now limited to local setting only)
- Composite action, which is planned to be used with Core repo as well, to run the full template on new builds of the Core
- Added the latest promoter for those who don't want to compare metrics

**What was changed**
- I revisited the materializer to make it a bit more clever
- Reworked documentation to fit better templating

**Note** Had to skip template runs and auto-formatting on Windows, due to the MLFlow deployer missing and OS-specific issues. We can revisit it later, if full Windows support will be there.

This PR also somewhat relates to [OSS-2135] - action out of this repo planned to be reused for testing in Core.

[OSS-2135]: https://zenml.atlassian.net/browse/OSS-2135?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,254 @@
...
+<p align="center">
+  <img height=400 src="assets/02_hp.png">
+</p>
+
+To ensure the high quality of ML models many ML Engineers go for automated hyperparameter tuning or even automated model architecture search. In this example, we are using prepared data from ETL to spin up a search of the best model parameters for different architectures in parallel.
+
+To create parallel processing of computationally expensive operations we use a loop over predefined potential architectures and respective parameters search grid and create one step for each candidate. After the steps are created we need to collect results (one best model per each search step) in a `hp_tuning_select_best_model` step to define the final winner and pass it to training. To ensure that collection goes smoothly and in full we use an `after` statement populated with all search steps names, so the selector job will wait for the completion of all searches.
+
+You can find more information about the current state of [Hyperparameter Tuning using ZenML in documentation](https://docs.zenml.io/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning).
```
Comment (strickvl) [inline README.md:115]:
```suggestion
You can find more information about the current state of [Hyperparameter Tuning using ZenML in the documentation](https://docs.zenml.io/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning).
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,254 @@
...
+
+
+### [Continuous Training] Training Pipeline: Model training and evaluation
+
+[📂 Code folder](template/steps/training/)
+<p align="center">
+  <img height=500 src="assets/03_train.png">
+</p>
+
+Having the best model architecture and its' hyperparameters defined in the previous stage makes it possible to train a quality model. Also, model training is the right place to bring an [Experiment Tracker](https://docs.zenml.io/user-guide/component-guide/experiment-trackers) into the picture - we will log all metrics and model itself into the [Experiment Tracker](https://docs.zenml.io/user-guide/component-guide/experiment-trackers), so we can register our model in a [Model Registry](https://docs.zenml.io/user-guide/component-guide/model-registries) and pass it down to a [Model Deployer](https://docs.zenml.io/user-guide/component-guide/model-deployers) easily and traceable. We will use information from Active Stack to make implementation agnostic of the underlying infrastructure.
```
Comment (strickvl) [inline README.md:140]:
```suggestion
Having the best model architecture and its hyperparameters defined in the previous stage makes it possible to train a quality model. Also, model training is the right place to bring an [Experiment Tracker](https://docs.zenml.io/user-guide/component-guide/experiment-trackers) into the picture - we will log all metrics and model itself into the [Experiment Tracker](https://docs.zenml.io/user-guide/component-guide/experiment-trackers), so we can register our model in a [Model Registry](https://docs.zenml.io/user-guide/component-guide/model-registries) and pass it down to a [Model Deployer](https://docs.zenml.io/user-guide/component-guide/model-deployers) easily and traceable. We will use information from Active Stack to make implementation agnostic of the underlying infrastructure.
```

File: template/Makefile [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+remote-login:
+	zenml connect --url "{{zenml_server_url}}"
+
+setup: remote-login
+{%- else %}
+setup:
+{%- endif %}
+	pip install -r requirements.txt
+	zenml integration install sklearn mlflow s3 kubernetes kubeflow slack evidently -y
+# TODO: add stack configuration ^^^
```
Comment (strickvl) [inline template/Makefile:11]: TODO

**Review by strickvl — CHANGES_REQUESTED** [review]
Some tiny things, but otherwise looks ok. There are some remaining pieces to [the bigger PRD](https://www.notion.so/zenml/ccf784a342d34531aeb80b6ab08e5e9d?v=b6142c3b538844929f76e9a364310898&p=a5072f0639bb4526a3c41a4e7d2a40ed&pm=s) but they'll need to be done separately / outside of this PR.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,254 @@
...
+
+It showcases the core ZenML concepts for supervised ML with batch predictions:
+
+* designing [ZenML pipeline steps](https://docs.zenml.io/user-guide/starter-guide/create-an-ml-pipeline)
+* using [step parameterization](https://docs.zenml.io/user-guide/starter-guide/create-an-ml-pipeline#parametrizing-a-step)
+ and [step caching](https://docs.zenml.io/user-guide/starter-guide/cache-previous-executions#caching-at-a-step-level)
+to design flexible and reusable steps
+* using [custom data types for your artifacts and writing materializers for them](https://docs.zenml.io/user-guide/advanced-guide/artifact-management/handle-custom-data-types)
+* constructing and running a [ZenML pipeline](https://docs.zenml.io/user-guide/starter-guide/create-an-ml-pipeline)
+* accessing ZenML pipeline run artifacts in [the post-execution phase](https://docs.zenml.io/user-guide/starter-guide/fetch-runs-after-execution)
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply README.md:35]: Links checked 👌🏼

File: template/Makefile (hunk 3 above)
Comment (avishniakov) [reply template/Makefile]: Added

File: .github/actions/e2e_template_test/action.yml [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      shell: bash
+      run: |
+        pip install --force --upgrade git+https://github.com/zenml-io/zenml.git@develop
+
+    - name: Run pytests
+      shell: bash
+      env:
+        ZENML_STACK_NAME: ${{ inputs.stack-name }}
+      run: |
+        pytest tests
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply .github/actions/e2e_template_test/action.yml:75]: Added

File: .github/actions/e2e_template_test/action.yml [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    
+    - name: Configure git (Windows)
+      if: ${{ inputs.os == 'windows-latest' }}
+      shell: bash
+      run: |
+        "C:\Program Files\Git\bin\git.exe" config --global user.email "info@zenml.io"
+        "C:\Program Files\Git\bin\git.exe" config --global user.name "ZenML GmbH"
+
+    # e2e example was not released yet, so relies on some recent develop changes
+    # TODO: replace with normal install after #1715, #1711, #1707 and #1703 are released
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply .github/actions/e2e_template_test/action.yml:35]: @strickvl , this is unresolvable as of now, due to the missing release of the above-mentioned PRs in the core. What do you do in such cases normally? Remove TODOs and spinn a ticket in Jira or?

File: .github/actions/e2e_template_test/action.yml [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        requirement_files: all-requirements.txt
+        
+    - name: Install requirements
+      if: steps.cache-virtualenv.outputs.cache-hit != 'true'
+      shell: bash
+      run: |
+        pip install -r all-requirements.txt
+    
+    # e2e example was not released yet, so relies on some recent develop changes
+    # TODO: remove step after #1715, #1711, #1707 and #1703 are released
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply .github/actions/e2e_template_test/action.yml:64]: @strickvl , this is unresolvable as of now, due to the missing release of the above-mentioned PRs in the core. What do you do in such cases normally? Remove TODOs and spinn a ticket in Jira or?

File: .github/actions/e2e_template_test/action.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (strickvl) [reply .github/actions/e2e_template_test/action.yml:35]: Leave TODOs, but yes add a JIRA ticket to OSS board.

File: .github/actions/e2e_template_test/action.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (strickvl) [reply .github/actions/e2e_template_test/action.yml:64]: Leave TODOs, but yes add a JIRA ticket to OSS board.

**Review by strickvl — APPROVED** [review]

**Pull request merged**